activerecord/lib/active_record/autosave_association.rb

```diff
@@ -390,11 +390,8 @@ def save_has_one_association(reflection)
           elsif autosave != false
             key = reflection.options[:primary_key] ? send(reflection.options[:primary_key]) : id
 
-            if (autosave && record.changed_for_autosave?) || new_record? || record_changed?(reflection, record, key)
-              unless reflection.through_reflection
-                record[reflection.foreign_key] = key
-              end
-
+            if (autosave && record.changed_for_autosave?) || (!reflection.through_reflection && record_changed?(reflection, record, key))
```

Comment:
I was just looking back at this. `new_record?`is redundant as its implied by the previous conditions. I think I was just trying to simplify.